### ISSUE TITLE
Speed up protein mutation charge changing repex consistency test

### ISSUE THREAD

**Issue by zhang-ivy** [choderalab/perses#1058]
This test is currently very slow, so we need to speed up. 

See: https://github.com/choderalab/perses/issues/1044#issuecomment-1165575916

**Comment by zhang-ivy** [comment]
I just analyzed the dipeptide charge changing test data: https://github.com/choderalab/perses/issues/1044#issuecomment-1167913833

**Comment by zhang-ivy** [comment]
@ijpulidos : This should speed the test up a lot:
```python
move = mcmc.LangevinDynamicsMove(timestep=4.0 * unit.femtoseconds,
                                                                       collision_rate=1.0 / unit.picosecond,
                                                                       n_steps=50,
                                                                       reassign_velocities=False,
                                                                       n_restart_attempts=20)
```

Note that we want to wait until openmmtools 0.21.5 is released so that we can use the updated `LangevinDynamicsMove` with the `LangevinMiddleIntegrator`

**Comment by ijpulidos** [comment]
By using `openmm-nightly` and `LangevinDynamicsMove` we are getting almost a 4x speedup for the neutral test:

![neutral_77-LangSplitDynMove_vs_nightly-LangDynMove](https://user-images.githubusercontent.com/2949729/185011365-566dfc9a-501d-45ce-9f80-55604415d0d5.png)

This means that for the neutral mutation test, for 1000 iterations, we are talking about ~35 minutes in a 1080Ti GPU (1 s per iteration, 1000 iterations and 2 replicates -- forward and backward --). Which should fit our GPU CI workflow.

For the charge mutation (explicit solvent) the speedup is not that high, we are getting about a 30% increase in performance

![charge_77-LangSplitDynMove_vs_nightly-LangDynMove](https://user-images.githubusercontent.com/2949729/185014788-f8ffaf94-6144-4025-b62c-7427c005a6bd.png)

that would mean the tests would take about ~9 hours for 1000 iterations on a 1080Ti GPU (8 s per iteration, 1000 iterations and 4 replicates as per the test goes)

**Comment by zhang-ivy** [comment]
Great! Also, I just checked and the [current tests](https://github.com/choderalab/perses/blob/main/perses/tests/test_repex.py) use 3000 iterations -- Why are you using 1000 iterations? Is the test still passing (i.e. `assert DDG < 6 * dDDG, f"DDG ({DDG}) is greater than 6 * dDDG ({6  * dDDG})"` doesnt raise an error), even with 1000 iterations?

**Comment by ijpulidos** [comment]
> Great! Also, I just checked and the [current tests](https://github.com/choderalab/perses/blob/main/perses/tests/test_repex.py) use 3000 iterations -- Why are you using 1000 iterations? Is the test still passing (i.e. `assert DDG < 6 * dDDG, f"DDG ({DDG}) is greater than 6 * dDDG ({6 * dDDG})"` doesnt raise an error), even with 1000 iterations?

I was thinking about 1000 iterations for the computations using your comment [here](https://github.com/choderalab/perses/issues/1044#issuecomment-1167913833). The test passed, but I just run it once, even though the numbers in that same thread seem to imply it might have issues with 1000 iterations. I guess we can bump it up to 2000 to be "safe" and it would still just be over 1 hour on the CI, which should be okay.

**Comment by zhang-ivy** [comment]
Ah that comment was referring to the charge changing mutations. For the neutral mutations, you'll have to see my comment [here](https://github.com/choderalab/perses/issues/1044#issuecomment-1165575916)

**Comment by zhang-ivy** [comment]
But both of those comments are for experiments using 50 steps per iteration, instead of 250 steps (which is what we normally use in our production runs). Are you using 250 steps per iteration, or 50 steps?

**Comment by ijpulidos** [comment]
> Ah that comment was referring to the charge changing mutations. For the neutral mutations, you'll have to see my comment [here](https://github.com/choderalab/perses/issues/1044#issuecomment-1165575916)

I see, well I guess we can use 3000 iterations and it would be 1.5-2 hours on a 1080Ti (we have to check our GPU CI has a similar performance but I suspect our GPU CI is currently slower, since it is using older Tesla K40). 

> But both of those comments are for experiments using 50 steps per iteration, instead of 250 steps (which is what we normally use in our production runs). Are you using 250 steps per iteration, or 50 steps?

50 steps, as the tests go.

```python
        hss = HybridRepexSampler(mcmc_moves=mcmc.LangevinSplittingDynamicsMove(timestep=4.0 * unit.femtoseconds,
                                                                        collision_rate=1.0 / unit.picosecond,
                                                                        n_steps=50,
                                                                        reassign_velocities=False,
                                                                        n_restart_attempts=20,
                                                                        splitting="V R R R O R R R V",
                                                                        constraint_tolerance=1e-06),
                                    replica_mixing_scheme='swap-all',
                                    hybrid_factory=htf,
                                    online_analysis_interval=None)
```

**Comment by jchodera** [comment]
First of all, great news about the speedup from using `openmm-dev` and the new `LangevinMiddleIntegrator`.

I'm a bit surprised at why there is an 8x difference in iteration time between the charge and neutral tests. I think there are more thermodynamic states in the charge test, but any ideas what might lead to the 8x difference in iteration time?

**Comment by zhang-ivy** [comment]
>I see, well I guess we can use 3000 iterations and it would be 1.5-2 hours on a 1080Ti (we have to check our GPU CI has a similar performance but I suspect our GPU CI is currently slower, since it is using older Tesla K40).

Sounds good!

>I'm a bit surprised at why there is an 8x difference in iteration time between the charge and neutral tests. I think there are more thermodynamic states in the charge test, but any ideas what might lead to the 8x difference in iteration time?

The neutral test uses 12 replicas, the charge test use 36 replicas. Additionally, the neutral test only requires two transformations (A->T, T->A), whereas charge test requires 4. I think the 8x difference makes sense given the extra thermodynamic states and transformations.

**Comment by zhang-ivy** [comment]
Oh.. I see the problem. Its an 8x difference per iteration (without considering the extra transformations)...

**Comment by zhang-ivy** [comment]
Interestingly, when comparing the neutral vs charge test iteration time with the LangevinSplittingDynamicsMove, and openmm 7.7, charge is in fact 3x slower than neutral.

For LangevinDynamicsMove + openmm nightly, I wonder if the GPUs being used for neutral vs charge are different? @ijpulidos Any other ideas why they are different?

**Comment by ijpulidos** [comment]
@zhang-ivy Ah good catch, I just checked and I did use a wrong node in one of the tests (I misread `ls-gpu` nodes on Lilac to be 1080ti and these are only 1080). This only affects the first plot (comparison between neutral tests). I'll re-run this test on the correct node and update the plot once I get the results.

Comparing neutral vs charge, we have to consider that the tests currently use 12 replicas for neutral and 36 for charge (which also uses explicit solvent), that may help explaining the differences in iteration times.

**Comment by jchodera** [comment]
> Comparing neutral vs charge, we have to consider that the tests currently use 12 replicas for neutral and 36 for charge (which also uses explicit solvent), that may help explaining the differences in iteration times.

This would be responsible for a 3x difference, but it's hard to imagine the 1080Ti vs 1080 is responsible for the remaining remaining ~2.7x iteration time difference.

I've made a few comments on https://github.com/choderalab/perses/pull/1065 to suggest some ways we might be able to speed things up a bit.

**Comment by ijpulidos** [comment]
> This would be responsible for a 3x difference, but it's hard to imagine the 1080Ti vs 1080 is responsible for the remaining remaining ~2.7x iteration time difference.

Yeah, that doesn't explain the remaining difference, I just updated the results.

> I've made a few comments on #1065 to suggest some ways we might be able to speed things up a bit.

Will take a look at those and let you know once I get the chance to test them. Thanks for taking the time to look into this!

**Comment by zhang-ivy** [comment]
>I've made a few comments on https://github.com/choderalab/perses/pull/1065 to suggest some ways we might be able to speed things up a bit.

@jchodera : Thanks for the feedback! I just responded to all your comments. 

>Yeah, that doesn't explain the remaining difference

I just double checked my perses barnase:barstar paper data -- I ran A->T dipeptide and R->A dipeptide,  both in solvent, using 2 gpus, and am seeing that A->T takes ~0.68 sec per iteration and R->A takes ~1.5 sec per iteration  on lu nodes (GeForce RTX 2080). 
@ijpulidos : Can you try re-running all the tests? And if you are still seeing an 8-fold difference, let's set up a quick chat to discuss what might be causing the difference as I am not seeing the same slowdown in my data.

**Comment by ijpulidos** [comment]
> @ijpulidos : Can you try re-running all the tests? And if you are still seeing an 8-fold difference, let's set up a quick chat to discuss what might be causing the difference as I am not seeing the same slowdown in my data.

Please refer to my previous comment for the updated information https://github.com/choderalab/perses/issues/1058#issuecomment-1217359114 . I settled on using my local A4000 GPU due to some replicability issues when using lilac 1080ti nodes.

I re-run the tests and updated the plots. Basically we were facing the problem that @zhang-ivy has already warned us about in the last part of https://github.com/choderalab/perses/pull/1054#issuecomment-1178224781

We have a slowdown problem when we generate the HTF and run the simulation in the same script, this was causing the extra overhead. To avoid this I had to save the HTFs to pickle files and then load them to run the simulation in a separate script. We don't know why this is happening but we will try to debug this further on some other issue report.

Basically we can now verify the consistency on the updated plots where you get a ~4x iteration time difference when comparing the two tests (that is comparing between the same colored lines accross the two plots). We need to refactor the tests to account for things discussed in this thread.

**Comment by zhang-ivy** [comment]
To address the slowdown problem, we don't actually want to run separate scripts for the test, so Iván and I discussed that the next short term action item is to see if generating the htf, saving the htf as a pickle file, deleting the `htf`, and then loading it in from the pickle file prevents the slowdown -- if it does, we can adapt the tests to do this. And then we can open an issue to investigate this issue further in the long term.

**Comment by jchodera** [comment]
Wow, this is bizarre. Can someone point me to the code that sets up the HybridTopologyFactory that exhibits this weird behavior? I wonder if it's leaving the ContextCaches in a weird state that might be forcing them to use the CPU platform some of the time.

> To address the slowdown problem, we don't actually want to run separate scripts for the test, so Iván and I discussed that the next short term action item is to see if generating the htf, saving the htf as a pickle file, deleting the htf, and then loading it in from the pickle file prevents the slowdown -- if it does, we can adapt the tests to do this. And then we can open an issue to investigate this issue further in the long term.

This sounds like a reasonable approach to get this out the door---we can fix this with #1084.

**Comment by zhang-ivy** [comment]
> Can someone point me to the code that sets up the HybridTopologyFactory that exhibits this weird behavior? 

Here's the neutral test: https://github.com/choderalab/perses/blob/main/perses/tests/test_repex.py#L6-L84 
You should be able to reproduce the problem with this test.

**Comment by jchodera** [comment]
@zhang-ivy : I just tried running that test, but it doesn't seem to actually run a simulation:
```bash
$ pytest --verbose --log-cli-level=DEBUG perses/tests/test_repex.py::test_RESTCapableHybridTopologyFactory_repex_neutral_mutation
============================================================================================================================== test session starts ==============================================================================================================================
platform linux -- Python 3.9.12, pytest-7.1.2, pluggy-1.0.0 -- /lila/home/chodera/miniconda/envs/perses/bin/python
cachedir: .pytest_cache
rootdir: /lila/home/chodera/github/choderalab/perses/perses
collecting ... 
------------------------------------------------------------------------------------------------------------------------------ live log collection ------------------------------------------------------------------------------------------------------------------------------
WARNING  root:__init__.py:5 Warning: importing 'simtk.openmm' is deprecated.  Import 'openmm' instead.
INFO     numexpr.utils:utils.py:144 Note: detected 128 virtual cores but NumExpr set to maximum of 64, check "NUMEXPR_MAX_THREADS" environment variable.
INFO     numexpr.utils:utils.py:147 Note: NumExpr detected 128 cores but "NUMEXPR_MAX_THREADS" not set, so enforcing safe limit of 8.
INFO     numexpr.utils:utils.py:159 NumExpr defaulting to 8 threads.
collected 1 item                                                                                                                                                                                                                                                                

perses/tests/test_repex.py::test_RESTCapableHybridTopologyFactory_repex_neutral_mutation 
--------------------------------------------------------------------------------------------------------------------------------- live log call ---------------------------------------------------------------------------------------------------------------------------------
WARNING  openff.toolkit.utils.toolkit_registry:toolkit_registry.py:189 Warning: Unable to load toolkit 'OpenEye Toolkit'. The Open Force Field Toolkit does not require the OpenEye Toolkits, and can use RDKit/AmberTools instead. However, if you have a valid license for the OpenEye Toolkits, consider installing them for faster performance and additional file format support: https://docs.eyesopen.com/toolkits/python/quickstart-python/linuxosx.html OpenEye offers free Toolkit licenses for academics: https://www.eyesopen.com/academic-licensing
INFO     setup:relative_point_mutation_setup.py:660 solvating at 0.15 M using tip3p
INFO     proposal_generator:topology_proposal.py:560 	Conducting polymer point mutation proposal...
INFO     proposal_generator:topology_proposal.py:1172 local_atom_map: {6: 6, 7: 7, 8: 8, 9: 9, 10: 10, 11: 13, 12: 14}
INFO     proposal_generator:topology_proposal.py:1175 the mapped atom names are: [('N', 'N'), ('H', 'H'), ('CA', 'CA'), ('HA', 'HA'), ('CB', 'CB'), ('C', 'C'), ('O', 'O')]
$
```
Am I doing something wrong? This only takes a few seconds.

**Comment by jchodera** [comment]
OK, I'm running this on an A40, and have noticed a few things:
1. The GPU is definitely in use during the setup phase. This is likely using one or more `ContextCache`s
2. After `hss.setup()`, we then swap this out with two new _empty_ `ContextCache`s without cleaning up the old one
3. When it runs, it hits a ton of `NaN`s and has to keep doing integrator restarts, which will slow things down a lot:
```
WARNING:openmmtools.mcmc:Potential energy is NaN after 1 attempts of integration with move LangevinSplittingDynamicsMove Attempting a restart...
WARNING:openmmtools.mcmc:Potential energy is NaN after 0 attempts of integration with move LangevinSplittingDynamicsMove Attempting a restart...
WARNING:openmmtools.mcmc:Potential energy is NaN after 0 attempts of integration with move LangevinSplittingDynamicsMove Attempting a restart...
WARNING:openmmtools.mcmc:Potential energy is NaN after 0 attempts of integration with move LangevinSplittingDynamicsMove Attempting a restart...
WARNING:openmmtools.mcmc:Potential energy is NaN after 0 attempts of integration with move LangevinSplittingDynamicsMove Attempting a restart...
WARNING:openmmtools.mcmc:Potential energy is NaN after 1 attempts of integration with move LangevinSplittingDynamicsMove Attempting a restart...
WARNING:openmmtools.mcmc:Potential energy is NaN after 0 attempts of integration with move LangevinSplittingDynamicsMove Attempting a restart...
WARNING:openmmtools.mcmc:Potential energy is NaN after 0 attempts of integration with move LangevinSplittingDynamicsMove Attempting a restart...
WARNING:openmmtools.mcmc:Potential energy is NaN after 0 attempts of integration with move LangevinSplittingDynamicsMove Attempting a restart...
WARNING:openmmtools.mcmc:Potential energy is NaN after 1 attempts of integration with move LangevinSplittingDynamicsMove Attempting a restart...
WARNING:openmmtools.mcmc:Potential energy is NaN after 0 attempts of integration with move LangevinSplittingDynamicsMove Attempting a restart...
WARNING:openmmtools.mcmc:Potential energy is NaN after 1 attempts of integration with move LangevinSplittingDynamicsMove Attempting a restart...
WARNING:openmmtools.mcmc:Potential energy is NaN after 0 attempts of integration with move LangevinSplittingDynamicsMove Attempting a restart...
WARNING:openmmtools.mcmc:Potential energy is NaN after 0 attempts of integration with move LangevinSplittingDynamicsMove Attempting a restart...
WARNING:openmmtools.mcmc:Potential energy is NaN after 1 attempts of integration with move LangevinSplittingDynamicsMove Attempting a restart...
WARNING:openmmtools.mcmc:Potential energy is NaN after 0 attempts of integration with move LangevinSplittingDynamicsMove Attempting a restart...
...
```

**Comment by jchodera** [comment]
@ijpulidos: Can you remind me how I can get `pytest` to have this test generate per-iteration verbose output of timing information? I'm not having any luck with combinations like
```bash
$ pytest -s --verbose --log-level=DEBUG --log-cli-level=DEBUG perses/tests/test_repex.py::test_RESTCapableHybridTopologyFactory_repex_neutral_mutation
```

**Comment by jchodera** [comment]
I managed to get the iteration info to print by running the test as `__main__` and inserting
```python

```
The current `main` version runs each iteration at ~1.4 s/iteration for the first phase and ~1.4 s/iteration for the second phase.

However, I don't have something to compare it to where I serialize out the state and reload it in a different process to see it get faster by >2x. Do you have an example of that?

**Comment by zhang-ivy** [comment]
>The GPU is definitely in use during the setup phase. This is likely using one or more ContextCaches

Yes, Iván and I discussed this yesterday -- we are wondering whether its the geometry engine, which doesn't actually use a ContextCache, just a context: https://github.com/choderalab/perses/blob/main/perses/rjmc/geometry.py#L453-L463

The only other place that uses contexts in the `PointMutationExecutor` pipeline is [when we validate endstate energies](https://github.com/choderalab/perses/blob/main/perses/app/relative_point_mutation_setup.py#L437-L473), but that part is [disabled for the test](https://github.com/choderalab/perses/blob/main/perses/tests/test_repex.py#L42), so I think its the geometry engine

**Comment by zhang-ivy** [comment]
>However, I don't have something to compare it to where I serialize out the state and reload it in a different process to see it get faster by >2x. Do you have an example of that?

You can generate the htf using: https://github.com/choderalab/perses/blob/main/perses/tests/test_repex.py#L34-L44
Then save a pickled htf using:
```python
with open(os.path.join(outdir, f"name.pickle"), "wb") as f:
        pickle.dump(htf, f)
```
Then run the test, skipping the htf generation and instead loading in the pickled htf using: 
```python
htf = pickle.load(open(os.path.join(outdir, f"name.pickle"), "rb" ))
```
I don't have an example pickled htf, but maybe Iván does.

**Comment by ijpulidos** [comment]
> However, I don't have something to compare it to where I serialize out the state and reload it in a different process to see it get faster by >2x. Do you have an example of that?

In the attached tar.gz file you can find a bash script to run the tests. It basically just calls the python script to generate and serialize the HTFs and then a separate script to run the repex simulation. You can also find the pickled topology proposals for Openmm 7.7.

[neutral_mutation_repex_test.tar.gz](https://github.com/choderalab/perses/files/9375381/neutral_mutation_repex_test.tar.gz)

**Comment by ijpulidos** [comment]
We need to double check, this was probably solved when we merged the repex tests.

**Comment by zhang-ivy** [comment]
I think we already sped things up a bit using this PR: https://github.com/choderalab/perses/pull/1065
There may be room for more speed up, but I think its fine for now. What do you think? @ijpulidos @mikemhenry
